### ISSUE TITLE
How to link yml and android project?(如何链接yml和android项目？)

### ISSUE THREAD

**Issue by chartores** [XiaoMi/mace#691]
Hello.
To use the current mace library, the settings and tests have been completed in basic PC and Android.

Through the process below, I have a rough understanding of how and how to create an application using the mace library on Android.

train.py -> xxx.meta / xxx.index / xxx.data -> gen_frozen_pb .py -> xxx.pb -> [Mace Library]converter.py -> xxx.yml

However, even if I look at the'mace' documentation, I don't understand the process of how to do it when connecting yml and Android.

I would really appreciate it if you could tell me how to do it in detail.

----------------------------------------------------------------------------------------------------------------------------------------------
你好。
为了使用当前的Mace库，已经在基本PC和Android上完成了设置和测试。

通过以下过程，我对如何以及如何使用Android上的Mace库创建应用程序有了一个粗略的了解。

train.py-> xxx.meta / xxx.index / xxx.data-> gen_frozen_pb .py-> xxx.pb-> [Mace库] converter.py-> xxx.yml

但是，即使我看了'mace'文档，也无法理解在连接yml和Android时如何进行操作的过程。

如果您能告诉我详细操作方法，我将不胜感激。

**Comment by lu229** [comment]
@chartores yml只是转换模型时使用的配置文件，转换完成获得Mace库之后，可以将Mace库的so或者.a文件（[文档](https://mace.readthedocs.io/en/latest/user_guide/basic_usage.html#deploy-your-model-into-applications)），把Mace库链接到你的android so中，然后上层java通过jni调用so，就可以了，jni的具体例子可以参考：https://github.com/XiaoMi/mace/tree/master/examples/android

**Comment by chartores** [comment]
@lu229 

感谢您详细解释。 理论部分是有意义的。 如果您看下面的过程，'*?????????????? *' 我不了解链接这些部分的部分。

train.py-> xxx.meta / xxx.index / xxx.data-> gen_frozen_pb .py-> xxx.pb-> 
->[Mace library] converter.py-> xxx.yml -> *?????????????? * -> xxx.so OR xxx.a -> jni -> 안드로이드(자바)

这是一个基本的猜测，但这是一种用“ xxx.a OR xxx.so”替换“ yml（训练后的pb文件配置文件）”并将其放入Android项目的方法吗？

**Comment by lu229** [comment]
@chartores Could you speak English? I can't understand what you mean in Chinese.....

**Comment by chartores** [comment]
@lu229  

Ah, I thought I was a Chinese national, so I translated it in Chinese... It seems that there were a lot of errors in the automatic translation. If you would like to write a comment in English again, thank you for the detailed explanation. (Actually, I am not good at both Chinese and English.)

Thanks to "lu229", little by little, we are looking for clues and solving them. Thank you.
But I still don't know if the android connection method in yml is the way I think it is.

If you look below, this is the process I think.

train.py-> xxx.meta / xxx.index / xxx.data->

gen_frozen_pb .py-> xxx.pb->

converter.py-> xxx.yml ->

?????????????? -> xxx.so OR xxx.a -> jni -> Android (Java)

Question 1: Considering the above process, it's my personal guess, but do you replace "xxx.yml" with "xxx.a" or xxx.so file and put it in your Android project??

Question 2: Do you replace "xxx.yml" with "xxx.a" or xxx.so file according to the document addressed in the previous comment?

Question 3: I am wondering if it is correct to change the lower part of the contents of the CmakeLists.txt file in the macelibrary folder in the project.
[set(mobilenet_lib ${CMAKE_SOURCE_DIR}/src/main/cpp/model/arm64-v8a/mobilenet.a)]

Actually, this process and method are very new and difficult, but thank you for your help every time.

**Comment by lu229** [comment]
@chartores 
Perhaps you should read the [document](https://mace.readthedocs.io/en/latest/user_guide/basic_usage.html) first.

Question 1: Considering the above process, it's my personal guess, but do you replace "xxx.yml" with "xxx.a" or xxx.so file and put it in your Android project??
   The above process (you thought) is not right, in MACE, we use `converter.py` to convert the model files(such as xxx.pb) to mace model file(or lib), the  xxx.yml  is a config file which is used to config the model's info for the `converter.py`. after the convert, you can get the model file(or lib). please refer to this [document](https://mace.readthedocs.io/en/latest/user_guide/basic_usage.html).

Question 2: Do you replace "xxx.yml" with "xxx.a" or xxx.so file according to the document addressed in the previous comment?
    No, please refer to this [document](https://mace.readthedocs.io/en/latest/user_guide/basic_usage.html) and build the "xxx.a" or xxx.so file.

Question 3: I am wondering if it is correct to change the lower part of the contents of the CmakeLists.txt file in the macelibrary folder in the project.
[set(mobilenet_lib ${CMAKE_SOURCE_DIR}/src/main/cpp/model/arm64-v8a/mobilenet.a)]

yes, if you convert your model (in code mode) success and get the model lib(.a), then you can modify this part to replace the `mobilenet.a` with your own model lib.

**Comment by chartores** [comment]
@lu229 
Wow... That's amazing. Thank you very much.
I never thought of converting a deep learning model trained with "converter.py" to "xxx.yml" and "xxx.lib".
Can I change the contents of the "CmakeLists.txt" file and upload the "xxx.yml" and "xxx.lib" files to the Android project?
Maybe it's right!!!
Thank you very much. It was a great help.
I hope you will try it today and next week and succeed.
Have a happy day.

**Comment by lu229** [comment]
@chartores 
you can convert the deep learning model trained with "converter.py" to "xxx.lib"  but not  "xxx.yml",  "xxx.yml" is only a config file, please read the [document](https://mace.readthedocs.io/en/latest/user_guide/basic_usage.html) first, or you can not try it successfully.

**Comment by chartores** [comment]
@lu229 
I checked again. I was mistaken.
"xxx.yml" -> "xxx.lib" is created.
Thank you.

**Comment by chartores** [comment]
@lu229
I checked again. I was mistaken.
"xxx.yml" -> "xxx.lib" is created.
Thank you.
I created an xxx.a file using converter.py and xxx.xml files.

**Comment by chartores** [comment]
@lu229 
We proceeded as lu229 informed us, and then proceeded to the next process.
There is already an xxx.a file (70MB) in the project in the link below.
The newly created xxx.a file is (38MB).
So, the old one was deleted and the new one was put in.

The test was completed by building in the previous project.
(https://github.com/edvardHua/PoseEstimationForMobile/tree/master/android_demo/demo_mace)
(https://github.com/edvardHua/PoseEstimationForMobile/tree/master/android_demo/demo_mace/macelibrary/src/main/cpp/lib/armeabi-v7a/cpm.a)

However, when I tried to build in Android Studio by replacing the xxx.a file, the following error occurred.
Question 1: Was the replacement method wrong? Or was it wrong to build in Android Studio?
Q2: Should I build with ABD?
Question 3: The size difference between the existing file and the new file is very different. What is the problem?
(When you read the above project, it seems that mace 0.9 was used. The newly created file used the latest version of mace.)
There are many difficulties.

///////////////////////////////////////////////////////////////////////////////////////////////////
Build command failed.
Error while executing process /home/peng_ai/android/cmake/3.6.4111459/bin/cmake with arguments {--build /home/peng_ai/PoseEstimationForMobile/android_demo/demo_mace/macelibrary/.externalNativeBuild/cmake/debug/armeabi-v7a --target mace_mobile_jni}
[1/1] Linking CXX shared library ../../../../build/intermediates/cmake/debug/obj/armeabi-v7a/libmace_mobile_jni.so
FAILED: : && /opt/android-ndk-r16b/toolchains/llvm/prebuilt/linux-x86_64/bin/clang++  --target=armv7-none-linux-androideabi --gcc-toolchain=/opt/android-ndk-r16b/toolchains/arm-linux-androideabi-4.9/prebuilt/linux-x86_64 --sysroot=/opt/android-ndk-r16b/sysroot -fPIC -isystem /opt/android-ndk-r16b/sysroot/usr/include/arm-linux-androideabi -D__ANDROID_API__=21 -g -DANDROID -ffunction-sections -funwind-tables -fstack-protector-strong -no-canonical-prefixes -march=armv7-a -mfloat-abi=softfp -mfpu=vfpv3-d16 -fno-integrated-as -mthumb -Wa,--noexecstack -Wformat -Werror=format-security  -std=c++11 -fopenmp -O0 -fno-limit-debug-info  -Wl,--exclude-libs,libgcc.a -Wl,--exclude-libs,libatomic.a --sysroot /opt/android-ndk-r16b/platforms/android-21/arch-arm -Wl,--build-id -Wl,--warn-shared-textrel -Wl,--fatal-warnings -Wl,--fix-cortex-a8 -Wl,--no-undefined -Wl,-z,noexecstack -Qunused-arguments -Wl,-z,relro -Wl,-z,now -shared -Wl,-soname,libmace_mobile_jni.so -o ../../../../build/intermediates/cmake/debug/obj/armeabi-v7a/libmace_mobile_jni.so CMakeFiles/mace_mobile_jni.dir/src/main/cpp/spe.cc.o  ../../../../src/main/cpp/lib/armeabi-v7a/libmace.a ../../../../src/main/cpp/model/armeabi-v7a/cpm.a /opt/android-ndk-r16b/platforms/android-21/arch-arm/usr/lib/liblog.so -latomic -lm "/opt/android-ndk-r16b/sources/cxx-stl/gnu-libstdc++/4.9/libs/armeabi-v7a/libgnustl_static.a" && :
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::Argument* google::protobuf::Arena::CreateMaybeMessage<mace::Argument>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::Argument* google::protobuf::Arena::CreateMaybeMessage<mace::Argument>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::Argument* google::protobuf::Arena::CreateMaybeMessage<mace::Argument>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::InputOutputInfo* google::protobuf::Arena::CreateMaybeMessage<mace::InputOutputInfo>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::InputOutputInfo* google::protobuf::Arena::CreateMaybeMessage<mace::InputOutputInfo>(google::protobuf::Arena*)'
bazel-out/armeabi-v7a-opt/bin/include/_virtual_includes/public_headers/mace/port/logger.h:51: error: undefined reference to 'mace::port::VLogLevelFromStr(char const*)'
bazel-out/armeabi-v7a-opt/bin/include/_virtual_includes/public_headers/mace/port/logger.h:45: error: undefined reference to 'mace::port::LogLevelFromStr(char const*)'
bazel-out/armeabi-v7a-opt/bin/include/_virtual_includes/public_headers/mace/port/env.h:66: error: undefined reference to 'mace::port::Env::Default()'
bazel-out/armeabi-v7a-opt/bin/include/_virtual_includes/public_headers/mace/utils/logging.h:103: error: undefined reference to 'mace::port::Logger::Logger(char const*, int, mace::LogLevel)'
bazel-out/armeabi-v7a-opt/bin/include/_virtual_includes/public_headers/mace/utils/logging.h:103: error: undefined reference to 'mace::port::Logger::~Logger()'
bazel-out/armeabi-v7a-opt/bin/include/_virtual_includes/public_headers/mace/utils/logging.h:103: error: undefined reference to 'mace::port::Logger::~Logger()'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::OperatorDef* google::protobuf::Arena::CreateMaybeMessage<mace::OperatorDef>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::OperatorDef* google::protobuf::Arena::CreateMaybeMessage<mace::OperatorDef>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::OperatorDef* google::protobuf::Arena::CreateMaybeMessage<mace::OperatorDef>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::OperatorDef* google::protobuf::Arena::CreateMaybeMessage<mace::OperatorDef>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::ConstTensor* google::protobuf::Arena::CreateMaybeMessage<mace::ConstTensor>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::ConstTensor* google::protobuf::Arena::CreateMaybeMessage<mace::ConstTensor>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::ConstTensor* google::protobuf::Arena::CreateMaybeMessage<mace::ConstTensor>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::ConstTensor* google::protobuf::Arena::CreateMaybeMessage<mace::ConstTensor>(google::protobuf::Arena*)'
bazel-out/armeabi-v7a-opt/bin/include/_virtual_includes/public_headers/mace/port/env.h:66: error: undefined reference to 'mace::port::Env::Default()'
bazel-out/armeabi-v7a-opt/bin/include/_virtual_includes/public_headers/mace/utils/logging.h:109: error: undefined reference to 'mace::port::Logger::Logger(char const*, int, mace::LogLevel)'
bazel-out/armeabi-v7a-opt/bin/include/_virtual_includes/public_headers/mace/utils/logging.h:109: error: undefined reference to 'mace::port::Logger::~Logger()'
bazel-out/armeabi-v7a-opt/bin/include/_virtual_includes/public_headers/mace/utils/logging.h:109: error: undefined reference to 'mace::port::Logger::~Logger()'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::Argument* google::protobuf::Arena::CreateMaybeMessage<mace::Argument>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::OutputShape* google::protobuf::Arena::CreateMaybeMessage<mace::OutputShape>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::OutputShape* google::protobuf::Arena::CreateMaybeMessage<mace::OutputShape>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::OutputShape* google::protobuf::Arena::CreateMaybeMessage<mace::OutputShape>(google::protobuf::Arena*)'
external/com_google_protobuf/src/google/protobuf/repeated_field.h:643: error: undefined reference to 'mace::OutputShape* google::protobuf::Arena::CreateMaybeMessage<mace::OutputShape>(google::protobuf::Arena*)'
clang++: error: linker command failed with exit code 1 (use -v to see invocation)
ninja: build stopped: subcommand failed.
///////////////////////////////////////////////////////////////////////////////////////////////////

**Comment by lu229** [comment]
@chartores 
A1: It is right, but perhaps the lib lost something
A2: no need
A3: the lib perhaps lost something.

Please upload your `yml` file and tell me what is the command you used, so  I can check what happened.

**Comment by chartores** [comment]
@lu229 
The error was a mace version compatibility issue. (ver.latest -> ver.0.9)
No error on build.
However, there is no problem when building an Android application, but a problem occurs when running.
Perhaps even with the same version 0.9, there was a slight time difference.
(The reason for using mace-0.9 is that the existing project was made with version 0.9, so it was impossible to build an Android project with other versions. [In addition, Android apps like PC are divided into 32bit and 64bit, but only 32bit apps are available. ])
-Part 1--------------------------------------------------------------------------------------------------------
So I am trying to make libmace.a and xxx.a on mace 0.9 version.
I tried to build a new project on my computer at the address below.
(https://github.com/edvardHua/PoseEstimationForMobile/tree/master/android_demo/demo_mace)
And I linked the built Android project to the following address.
(https://drive.google.com/file/d/1hctLB2I4_XWe2XViPuqn4x0ATZb1Y_4F/view?usp=sharing)
The Mace related files required for the Android project are listed as follows. Please check if it is correct. haha
path: /demo_mace/macelibrary/src/main/cpp/include/mace/public
file: cpm_v1.h / mace.h / mace_engine_factory.h / mace_runtime.h
path: /demo_mace/macelibrary/src/main/cpp/lib/armeabi-v7a
file: libmace.a
path: /demo_mace/macelibrary/src/main/cpp/model/armeabi-v7a
file: cpm.a
-Part 2--------------------------------------------------------------------------------------------------------
And an error occurred while creating libmace.a.
Through the mace document, in the mace folder, proceeded as follows.
(Command: bash tools/build-standalone-lib.sh)
The environmental requirements were built exactly as for the document.
```
(mace-0.9) peng_ai@peng:~/mace-0.9$ bash tools/build-standalone-lib.sh
build shared lib for armeabi-v7a + cpu_gpu_dsp
Starting local Bazel server and connecting to it...
.............
INFO: Analysed target //mace/libmace:libmace_dynamic (26 packages loaded).
INFO: Found 1 target...
INFO: From Executing genrule @opencl_clhpp//:gen_opencl_clhpp:
/tmp/opencl-clhpp-build.Ju7VPCeyrC /home/peng_ai/.cache/bazel/_bazel_peng_ai/1b34a4d513700e8b18faf97011f679a9/sandbox/1070058944896470974/execroot/mace
/tmp/opencl-clhpp-build.Ju7VPCeyrC/build /tmp/opencl-clhpp-build.Ju7VPCeyrC /home/peng_ai/.cache/bazel/_bazel_peng_ai/1b34a4d513700e8b18faf97011f679a9/sandbox/1070058944896470974/execroot/mace
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/opencl-clhpp-build.Ju7VPCeyrC/build
Scanning dependencies of target generate_clhpp
[100%] Rebuilding cl.hpp ...
[100%] Built target generate_clhpp
Scanning dependencies of target generate_cl2hpp
[100%] Rebuilding cl2.hpp ...
[100%] Built target generate_cl2hpp
/tmp/opencl-clhpp-build.Ju7VPCeyrC /home/peng_ai/.cache/bazel/_bazel_peng_ai/1b34a4d513700e8b18faf97011f679a9/sandbox/1070058944896470974/execroot/mace
/home/peng_ai/.cache/bazel/_bazel_peng_ai/1b34a4d513700e8b18faf97011f679a9/sandbox/1070058944896470974/execroot/mace
installing to bazel-out/armeabi-v7a-opt/genfiles/external/opencl_clhpp
Target //mace/libmace:libmace_dynamic up-to-date (nothing to build)
INFO: Elapsed time: 72.933s, Critical Path: 13.01s
INFO: 115 processes, linux-sandbox.
INFO: Build completed successfully, 118 total actions
build shared lib for armeabi-v7a + cpu_gpu
INFO: Analysed target //mace/libmace:libmace_dynamic (0 packages loaded).
INFO: Found 1 target...
Target //mace/libmace:libmace_dynamic up-to-date (nothing to build)
INFO: Elapsed time: 49.048s, Critical Path: 10.26s
INFO: 99 processes, linux-sandbox.
INFO: Build completed successfully, 101 total actions
build shared lib for arm64-v8a + cpu_gpu
INFO: Analysed target //mace/libmace:libmace_dynamic (0 packages loaded).
INFO: Found 1 target...
INFO: From Executing genrule @opencl_clhpp//:gen_opencl_clhpp:
/tmp/opencl-clhpp-build.csWF4p0wHU /home/peng_ai/.cache/bazel/_bazel_peng_ai/1b34a4d513700e8b18faf97011f679a9/sandbox/5537821959544941460/execroot/mace
/tmp/opencl-clhpp-build.csWF4p0wHU/build /tmp/opencl-clhpp-build.csWF4p0wHU /home/peng_ai/.cache/bazel/_bazel_peng_ai/1b34a4d513700e8b18faf97011f679a9/sandbox/5537821959544941460/execroot/mace
-- The C compiler identification is GNU 5.4.0
-- The CXX compiler identification is GNU 5.4.0
-- Check for working C compiler: /usr/bin/cc
-- Check for working C compiler: /usr/bin/cc -- works
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Detecting C compile features
-- Detecting C compile features - done
-- Check for working CXX compiler: /usr/bin/c++
-- Check for working CXX compiler: /usr/bin/c++ -- works
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- Configuring done
-- Generating done
-- Build files have been written to: /tmp/opencl-clhpp-build.csWF4p0wHU/build
Scanning dependencies of target generate_clhpp
[100%] Rebuilding cl.hpp ...
[100%] Built target generate_clhpp
Scanning dependencies of target generate_cl2hpp
[100%] Rebuilding cl2.hpp ...
[100%] Built target generate_cl2hpp
/tmp/opencl-clhpp-build.csWF4p0wHU /home/peng_ai/.cache/bazel/_bazel_peng_ai/1b34a4d513700e8b18faf97011f679a9/sandbox/5537821959544941460/execroot/mace
/home/peng_ai/.cache/bazel/_bazel_peng_ai/1b34a4d513700e8b18faf97011f679a9/sandbox/5537821959544941460/execroot/mace
installing to bazel-out/arm64-v8a-opt/genfiles/external/opencl_clhpp
Target //mace/libmace:libmace_dynamic up-to-date (nothing to build)
INFO: Elapsed time: 54.137s, Critical Path: 12.15s
INFO: 140 processes, linux-sandbox.
INFO: Build completed successfully, 155 total actions
build shared lib for linux-x86-64
INFO: Analysed target //mace/libmace:libmace_dynamic (0 packages loaded).
INFO: Found 1 target...
ERROR: /home/peng_ai/mace-0.9/mace/core/BUILD:20:1: C++ compilation of rule '//mace/core:core' failed (Exit 1): gcc failed: error executing command 
  (cd /home/peng_ai/.cache/bazel/_bazel_peng_ai/1b34a4d513700e8b18faf97011f679a9/execroot/mace && \
  exec env - \
    PWD=/proc/self/cwd \
  /usr/local/cuda-10.0/bin/gcc -U_FORTIFY_SOURCE -fstack-protector -Wall -B/usr/local/cuda-10.0/bin -B/usr/bin -Wunused-but-set-parameter -Wno-free-nonheap-object -fno-omit-frame-pointer -g0 -O2 '-D_FORTIFY_SOURCE=1' -DNDEBUG -ffunction-sections -fdata-sections '-std=c++0x' -MD -MF bazel-out/k8-opt/bin/mace/core/_objs/core/mace/core/runtime/cpu/cpu_runtime.pic.d '-frandom-seed=bazel-out/k8-opt/bin/mace/core/_objs/core/mace/core/runtime/cpu/cpu_runtime.pic.o' -fPIC -iquote . -iquote bazel-out/k8-opt/genfiles -iquote external/bazel_tools -iquote bazel-out/k8-opt/genfiles/external/bazel_tools -iquote external/com_google_protobuf -iquote bazel-out/k8-opt/genfiles/external/com_google_protobuf -isystem external/com_google_protobuf/src -isystem bazel-out/k8-opt/genfiles/external/com_google_protobuf/src -isystem bazel-out/k8-opt/bin/external/com_google_protobuf/src '-std=c++11' -D_GLIBCXX_USE_C99_MATH_TR1 -DMACE_OBFUSCATE_LITERALS -O3 -ffunction-sections -fdata-sections '-fvisibility=hidden' -Werror -Wextra -Wno-missing-field-initializers -fopenmp -DMACE_ENABLE_OPENMP -fno-canonical-system-headers -Wno-builtin-macro-redefined '-D__DATE__="redacted"' '-D__TIMESTAMP__="redacted"' '-D__TIME__="redacted"' -c mace/core/runtime/cpu/cpu_runtime.cc -o bazel-out/k8-opt/bin/mace/core/_objs/core/mace/core/runtime/cpu/cpu_runtime.pic.o)

Use --sandbox_debug to see verbose messages from the sandbox
mace/core/runtime/cpu/cpu_runtime.cc: In function 'mace::MaceStatus mace::GetCPUBigLittleCoreIDs(std::vector<int>*, std::vector<int>*)':
mace/core/runtime/cpu/cpu_runtime.cc:95:12: error: '%d' directive output may be truncated writing between 1 and 10 bytes into a region of size 5 [-Werror=format-truncation=]
 MaceStatus GetCPUBigLittleCoreIDs(std::vector<int> *big_core_ids,
            ^~~~~~~~~~~~~~~~~~~~~~
mace/core/runtime/cpu/cpu_runtime.cc:95:12: note: directive argument in the range [0, 2147483647]
In file included from /usr/include/stdio.h:936:0,
                 from /usr/include/c++/7/cstdio:42,
                 from /usr/include/c++/7/ext/string_conversions.h:43,
                 from /usr/include/c++/7/bits/basic_string.h:6159,
                 from /usr/include/c++/7/string:52,
                 from /usr/include/c++/7/stdexcept:39,
                 from /usr/include/c++/7/array:39,
                 from /usr/include/c++/7/tuple:39,
                 from /usr/include/c++/7/bits/stl_map.h:63,
                 from /usr/include/c++/7/map:61,
                 from ./mace/public/mace.h:22,
                 from ./mace/core/runtime/cpu/cpu_runtime.h:20,
                 from mace/core/runtime/cpu/cpu_runtime.cc:15:
/usr/include/x86_64-linux-gnu/bits/stdio2.h:65:44: note: '__builtin_snprintf' output between 29 and 38 bytes into a destination of size 32
        __bos (__s), __fmt, __va_arg_pack ());
                                            ^
cc1plus: all warnings being treated as errors
Target //mace/libmace:libmace_dynamic failed to build
INFO: Elapsed time: 54.387s, Critical Path: 13.54s
INFO: 210 processes: 208 linux-sandbox, 2 local.
FAILED: Build did NOT complete successfully

```

**Comment by lu229** [comment]
@chartores It seems that this error is an old [bug](https://github.com/XiaoMi/mace/issues/111), the version 0.9 is too old that we don't maintain it now, perhaps you can use the latest version [1.0.0](https://github.com/XiaoMi/mace/releases/tag/v1.0.0)? which have much better performance than the old versions.

**Comment by chartores** [comment]
However, the latest version built the example Android project about two weeks ago with the following command, but an error occurred in the middle.
Are you really okay?
```
cd mace/examples/android
./build.sh dynamic
```
Or do I need to create a "libmace.a" file and update and build it?
`# if libmace.a is needed, update `macelibrary/CMakeLists.txt` and run with `./build.sh static`.`

**Comment by chartores** [comment]
@lu229 
I copied and modified the "build-standalone-lib.sh" file that creates the "libmace.a" file.

I checked the error and the "build-standalone-lib.sh" file and it creates a "libmace.a" file for various platforms.

So I deleted the code that creates the "libmace.a" file for platforms other than "build static lib for armeabi-v7a + cpu_gpu".

And I built the "libmace.a" file of "build static lib for armeabi-v7a + cpu_gpu".

Finally, the build was successful on mace-0.9.

Now just build training on other data and it will be complete.

wish me luck.
